pkg/ddl/table_mode_test.go

```diff
@@ -0,0 +1,272 @@
+	t *testing.T,
+	store kv.Storage,
+	de ddl.ExecutorForTest,
+	dbInfo *model.DBInfo,
+	tblInfo *model.TableInfo,
+	mode model.TableModeState,
+) error {
+	args := &model.AlterTableModeArgs{
+		TableMode: mode,
+		SchemaID:  dbInfo.ID,
+		TableID:   tblInfo.ID,
+	}
+	job := &model.Job{
+		Version:    model.JobVersion2,
+		SchemaID:   dbInfo.ID,
+		TableID:    tblInfo.ID,
+		Type:       model.ActionAlterTableMode,
+		BinlogInfo: &model.HistoryInfo{},
+		InvolvingSchemaInfo: []model.InvolvingSchemaInfo{
+			{
+				Database: dbInfo.Name.L,
+				Table:    tblInfo.Name.L,
+			},
+		},
+	}
+	ctx.SetValue(sessionctx.QueryString, "skip")
+	err := de.DoDDLJobWrapper(ctx, ddl.NewJobWrapperWithArgs(job, args, true))
+
+	if err == nil {
+		v := getSchemaVer(t, ctx)
+		checkHistoryJobArgs(t, ctx, job.ID, &historyJobArgs{ver: v, tbl: tblInfo})
+
+		testCheckTableState(t, store, dbInfo, tblInfo, model.StatePublic)
+		testCheckJobDone(t, store, job.ID, true)
+		checkTableModeTest(t, store, dbInfo, tblInfo, mode)
+	}
+
+	return err
+}
+
+func checkTableModeTest(t *testing.T, store kv.Storage, dbInfo *model.DBInfo, tblInfo *model.TableInfo, mode model.TableModeState) {
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnDDL)
+	err := kv.RunInNewTxn(ctx, store, false, func(ctx context.Context, txn kv.Transaction) error {
+		tt := meta.NewMutator(txn)
+		info, err := tt.GetTable(dbInfo.ID, tblInfo.ID)
+		require.NoError(t, err)
+		require.NotNil(t, info)
+		require.Equal(t, mode, info.TableMode)
+		return nil
+	})
+	require.NoError(t, err)
+}
+
+func checkErrorCode(t *testing.T, err error, expected int) {
+	originErr := errors.Cause(err)
+	tErr, ok := originErr.(*terror.Error)
+	require.True(t, ok)
+	sqlErr := terror.ToSQLError(tErr)
+	require.Equal(t, expected, int(sqlErr.Code))
+}
+
+func TestTableModeBasic(t *testing.T) {
+	store, domain := testkit.CreateMockStoreAndDomain(t)
+	de := domain.DDLExecutor()
+	tk := testkit.NewTestKit(t, store)
+
+	ctx := testkit.NewTestKit(t, store).Session()
+
+	// init test
+	tk.MustExec("use test")
+	tk.MustExec("create table t1(id int)")
+
+	// get cloned table info for creating new table t1_restore
+	tblInfo := getClonedTableInfoFromDomain(t, "test", "t1", domain)
+
+	// For testing create table as ModeRestore
+	tblInfo.Name = ast.NewCIStr("t1_restore")
+	tblInfo.TableMode = model.TableModeRestore
+	err := de.CreateTableWithInfo(tk.Session(), ast.NewCIStr("test"), tblInfo, nil, ddl.WithOnExist(ddl.OnExistIgnore))
+	require.NoError(t, err)
+	dbInfo, ok := domain.InfoSchema().SchemaByName(ast.NewCIStr("test"))
+	require.True(t, ok)
+	checkTableModeTest(t, store, dbInfo, tblInfo, model.TableModeRestore)
+
+	// For testing select is not allowed when table is in ModeImport
+	tk.MustGetErrCode("select * from t1_restore", errno.ErrProtectedTableMode)
+
+	// For testing insert is not allowed when table is in ModeImport
+	tk.MustGetErrCode("insert into t1_restore values(1)", errno.ErrProtectedTableMode)
```

Comment:
add for `update/delete`, and all modifying DDLs, such as modify column/add column/add index, drop table, truncate table, etc.